sycl/plugins/esimd_cpu/esimd_libcm.hpp

```diff
@@ -0,0 +1,359 @@
+        LocalSize, [&](const sycl::id<__Dims__> &LocalID) {
+          sycl::id<__Dims__> GlobalID =
+              GroupID * LocalSize + LocalID + GlobalOffset;
+          const sycl::item<__Dims__, /*Offset=*/true> GlobalItem =
+              IDBuilder::createItem<__Dims__, true>(GlobalSize, GlobalID,
+                                                    GlobalOffset);
+          const sycl::item<__Dims__, /*Offset=*/false> LocalItem =
+              IDBuilder::createItem<__Dims__, false>(LocalSize, LocalID);
+          const sycl::nd_item<__Dims__> NDItem =
+              IDBuilder::createNDItem<__Dims__>(GlobalItem, LocalItem, Group);
+          ArgVector.push_back(NDItem);
+        });
+    for (int I = 0; I < __Dims__; ++I) {
+      SpaceDim[I] = (uint32_t)LocalSize[I];
+    }
+    run();
+  }
+
+private:
+  template <class ArgT = KernelArgType>
+  typename std::enable_if<std::is_same<ArgT, sycl::id<1>>::value>::type run() {
+    struct LambdaWrapper_ID_1DIM *WrappedLambda_ID_1DIM =
+        makeWrapper_ID_1DIM(ArgVector, MKernel);
+
+    ESimdCPUKernel ESimdCPU((fptrVoid)invokeLambda_ID_1DIM, SpaceDim);
+
+    ESimdCPU.launchMT(sizeof(struct LambdaWrapper_ID_1DIM),
+                      WrappedLambda_ID_1DIM);
+
+    ArgVector.clear();
+
+    delete WrappedLambda_ID_1DIM;
+  }
+
+  template <class ArgT = KernelArgType>
+  typename std::enable_if<std::is_same<ArgT, sycl::id<2>>::value>::type run() {
+    struct LambdaWrapper_ID_2DIM *WrappedLambda_ID_2DIM =
+        makeWrapper_ID_2DIM(ArgVector, MKernel);
+
+    ESimdCPUKernel ESimdCPU((fptrVoid)invokeLambda_ID_2DIM, SpaceDim);
+
+    ESimdCPU.launchMT(sizeof(struct LambdaWrapper_ID_2DIM),
+                      WrappedLambda_ID_2DIM);
+
+    ArgVector.clear();
+
+    delete WrappedLambda_ID_2DIM;
+  }
+
+  template <class ArgT = KernelArgType>
+  typename std::enable_if<std::is_same<ArgT, sycl::id<3>>::value>::type run() {
+    struct LambdaWrapper_ID_3DIM *WrappedLambda_ID_3DIM =
+        makeWrapper_ID_3DIM(ArgVector, MKernel);
+
+    ESimdCPUKernel ESimdCPU((fptrVoid)invokeLambda_ID_3DIM, SpaceDim);
+
+    ESimdCPU.launchMT(sizeof(struct LambdaWrapper_ID_3DIM),
+                      WrappedLambda_ID_3DIM);
+
+    ArgVector.clear();
+
+    delete WrappedLambda_ID_3DIM;
+  }
+
+  template <class ArgT = KernelArgType>
+  typename std::enable_if<std::is_same<ArgT, sycl::item<1, false>>::value>::type
+  run() {
+    struct LambdaWrapper_ITEM_1DIM *WrappedLambda_ITEM_1DIM =
+        makeWrapper_ITEM_1DIM(ArgVector, MKernel);
+
+    ESimdCPUKernel ESimdCPU((fptrVoid)invokeLambda_ITEM_1DIM, SpaceDim);
+
+    ESimdCPU.launchMT(sizeof(struct LambdaWrapper_ITEM_1DIM),
+                      WrappedLambda_ITEM_1DIM);
+
+    ArgVector.clear();
+
+    delete WrappedLambda_ITEM_1DIM;
+  }
+
+  template <class ArgT = KernelArgType>
+  typename std::enable_if<std::is_same<ArgT, sycl::item<2, false>>::value>::type
+  run() {
+    struct LambdaWrapper_ITEM_2DIM *WrappedLambda_ITEM_2DIM =
```

Comment:
Suggest using std::unique_ptr to avoid manual `delete`.

Reply:
Can 'std::unique_ptr<>' type be casted into 'void*' type?

Reply:
Yes, you just do `(fptrVoid)invokeLambda_ITEM_2DIM->get()`

Reply:
`invokeLambda_ITEM_2DIM.get()` worked.